### ISSUE TITLE
Improve system app uninstallation with a hard delete

### ISSUE THREAD

**Issue by undergroundwires** [undergroundwires/privacy.sexy#260]
### Problem description

Microsoft marks system apps with a `NonRemovable tag, making it more challenging to uninstall.

privacy.sexy *soft* deletes system apps by relocating them and making them inaccessible to the system (see [`windows.yaml`](https://github.com/undergroundwires/privacy.sexy/blob/659fea7afcabcd0ea273cfdcc8c4bae190c126f3/src/application/collections/windows.yaml#L7326-L7375)).

### Proposed solution

A new approach to address this problem has been detailed in the original issue. The NonRemovable tag can be removed by marking apps as end-of-life, allowing Remove-AppPackage to function properly. This involves adding keys to the registry for the PackageFullName of the package under:

```
HKLM\SOFTWARE\Microsoft\Windows\CurrentVersion\Appx\AppxAllUserStore\EndOfLife\$SID
```

The Edge package and other related packages can be removed using this method. Additionally, after uninstallation, the corresponding registry key can be deleted.

To avoid Windows Updates reinstalling the packages, the apps can be marked as deprovisioned using this registry key [1]:

```
HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Appx\AppxAllUserStore\Deprovisioned\
```

### Alternatives considered

The current method of renaming/removing binaries without proper uninstallation have following concerns:

- It may not be sustainable in the long run, especially with Windows updates.
- The current method also stores "user state" which can have privacy implications.
- Deleted apps still shown as installed apps which creates confusion.

### Additional information

This issue is created after discussions in #236.

[1]: https://learn.microsoft.com/en-us/windows/application-management/remove-provisioned-apps-during-update#create-registry-keys-for-deprovisioned-apps

**Comment by Nordmannn** [comment]
Hello undergroundwires,
Microsoft Edge Browser ist indeed very hard to kill. I would like to share a different approach. First one would have to deactivate the autorun entry using either "autoruns" or the taskmanger. After using my suggestions the Edge Browser will not be updated anymore.  

using elevated Powershell create these two rules:
New-NetFirewallRule -DisplayName "_BlockEdgeBrowser" -Direction Outbound -Program 'C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe' -Action Block
New-NetFirewallRule -DisplayName "_BlockEdgeUpdater" -Direction Outbound -Program 'C:\Program Files (x86)\Microsoft\EdgeUpdate\MicrosoftEdgeUpdate.exe' -Action Block

Turn off Microsoft Edge using gpedit.msc:
Computer Configuration > Administrative Template > Windows Components > Microsoft Edge
Activate and set to disabled  > Prevent Microsoft Edge from launching and loading the Start Page
Activate and set to disabled > New Tab page when Windows starts and each time Microsoft Edge is closed

using regedit to turn automatic start off:
reg add "HKCU\Software\Policies\Microsoft\MicrosoftEdge\Main" /v "AllowPrelaunch " /t REG_DWORD /d "0" /f
reg add "HKLM\Software\Policies\Microsoft\MicrosoftEdge\TabPreloader" /v "PreventTabPreloading" /t REG_DWORD /d "1" /f

using elevated cmd as Admin:
Set-Service -Name edgeupdate -StartupType Disabled
Set-Service -Name edgeupdatem -StartupType Disabled
Disable-ScheduledTask -TaskName 'MicrosoftEdgeUpdateTaskMachineCore'
Disable-ScheduledTask -TaskName 'MicrosoftEdgeUpdateTaskMachineUA'

**Comment by undergroundwires** [comment]
I've been testing this. The revert code does no longer work. Reversibility is important especially for system apps as they usually break a lot of things and reverting those has been helping many. So I suggest following approach for the logic:

1. Compress application data/metadata that's needed for recovery before deleting application and store in something like `%APPDATA%\privacy.sexy\revert-data\apps\{ApplicationName}`.
2. On revert, use these folders to reinstall the application and recover data/metadata.

Does this make sense? I appreciate any feedback from the community if you have reflections. Ping @TheAndr0id.

**Comment by undergroundwires** [comment]
The revert code does not work anymore. It's important to be able to revert these changes, as uninstalling system services can break a lot of things and it has already helped many and will help many others to be able to revert uninstallation of system services.

**Comment by TheAndr0id** [comment]
I just did some basic poking of the revert code and there might be an issue on how the Get-AppxPackage is being called.  Take a look at the following:

```
Get-AppxPackage -AllUsers Microsoft.Windows.Photos


Name                   : Microsoft.Windows.Photos
Publisher              : CN=Microsoft Corporation, O=Microsoft Corporation, L=Redmond, S=Washington, C=US
Architecture           : X64
ResourceId             :
Version                : 2023.10070.17002.0
PackageFullName        : Microsoft.Windows.Photos_2023.10070.17002.0_x64__8wekyb3d8bbwe
InstallLocation        : C:\Program Files\WindowsApps\Microsoft.Windows.Photos_2023.10070.17002.0_x64__8wekyb3d8bbwe
IsFramework            : False
PackageFamilyName      : Microsoft.Windows.Photos_8wekyb3d8bbwe
PublisherId            : 8wekyb3d8bbwe
PackageUserInformation : {S-1-5-18 [S-1-5-18]: Staged}
IsResourcePackage      : False
IsBundle               : False
IsDevelopmentMode      : False
NonRemovable           : False
IsPartiallyStaged      : False
SignatureKind          : Store
Status                 : Modified, NeedsRemediation

Name                   : Microsoft.Windows.Photos
Publisher              : CN=Microsoft Corporation, O=Microsoft Corporation, L=Redmond, S=Washington, C=US
Architecture           : X64
ResourceId             :
Version                : 2023.11090.12017.0
PackageFullName        : Microsoft.Windows.Photos_2023.11090.12017.0_x64__8wekyb3d8bbwe
InstallLocation        : C:\Program Files\WindowsApps\Microsoft.Windows.Photos_2023.11090.12017.0_x64__8wekyb3d8bbwe
IsFramework            : False
PackageFamilyName      : Microsoft.Windows.Photos_8wekyb3d8bbwe
PublisherId            : 8wekyb3d8bbwe
PackageUserInformation : {S-1-5-18 [S-1-5-18]: Staged}
IsResourcePackage      : False
IsBundle               : False
IsDevelopmentMode      : False
NonRemovable           : False
IsPartiallyStaged      : False
SignatureKind          : Store
```

You'll notice that there are two entries, and when the next two steps:

```
$manifest = $package.InstallLocation + '\AppxManifest.xml'

echo $manifest
C:\Program Files\WindowsApps\Microsoft.Windows.Photos_2023.10070.17002.0_x64__8wekyb3d8bbwe
C:\Program Files\WindowsApps\Microsoft.Windows.Photos_2023.11090.12017.0_x64__8wekyb3d8bbwe
\AppxManifest.xml
```
This will create an invalid entry for Add-AppxPackage. (  `Add-AppxPackage -DisableDevelopmentMode -Register "$manifest"`)

I didn't test this as this machine is now a "production" system that I'm not willing to mess up.  If you have a VM you can test, I'd see what happens if you use a valid $manifest with Add-AppxPackage.   I'm guessing (like MS Photos) some of the software has multiple versions stuck in the depot and that's causing the $manifest variable to be messed up.  Products like Edge (I'm also guessing things like Skype as well) may require the downloading and install of their web based installer.  The app installed via the windows install is basically a stub that updates itself as soon as Edge is run.

I'm still seeing multiple posts from 2023 still showing this method is valid.

I can understand wanting to have a back out path, that's a *VERY* good thing, but sometimes people have to take responsibility for their decisions.  I wanted Edge off my system, and *if* I change my mind on that I'll take responsibility that it might require yet another reinstall to fix it....

**Comment by undergroundwires** [comment]
This is great finding @TheAndr0id. It shows a bug where the revert code fails if more than one package metadata is found for single package. I will update the logic so it can handle this as improvement.

Otherwise, I have more good news 🥳 I have found a working and much more reliable way to reinstall these apps.

I have been looking at `winget` that seems to be able to reinstall store apps. And meanwhile I was testing I found [this documentation](https://learn.microsoft.com/en-us/windows/package-manager/winget/) that gave an example installation command: `Add-AppxPackage -RegisterByFamilyName -MainPackage Microsoft.DesktopAppInstaller_8wekyb3d8bbwe`. And I see that this works fine! It works with Edge and all other apps to revert those.

It uses `{packageName}_{publisherId}` syntax. E.g., for Edge it would be `Add-AppxPackage -RegisterByFamilyName -MainPackage Microsoft.MicrosoftEdge_8wekyb3d8bbwe`. So we'll need to hardcode publisher IDs inside privacy.sexy.

So **planned changes**:

1. **Handle Multiple Package Metadata:** Address the bug you've pointed out. When multiple metadata entries exist for a package, the current logic fails. I'll improve the code to handle these scenarios.
2. **Use `RegisterByFamilyName` app reinstallation**: 
   1. I will update revert code to run `Add-AppxPackage -RegisterByFamilyName`  if package cannot be found or `Add-AppxPackage -DisableDevelopmentMode -Register "$manifest"` fails.
   2. I need to figure out publisherIds for all of the apps and hardcode them in the code so so privacy.sexy can run `-RegisterByFamilyName` on revert.
   3. Meanwhile I'm doing this, I will improve documentation for all apps.
3. **Modify File Renaming Script:**  I think we can still keep the renaming script but change it so it works even though `Get-AppxPackage` returns null (package is non-existing), so we can run it after a hard delete. Currently, the script depends on `Get-AppxPackage.InstallLocation`. However, `InstallLocation` can be any of the three paths:
   - C:\Windows\SystemApps\{PackageFamilyName}
   - C:\Windows\{AppName}
   - C:\Program Files\WindowsApps\{PackageName}
   To ensure resilience, I'll adjust the script to rename files directly within these directories, bypassing the need for Get-AppxPackage.
4. Deprovision packages to prevent Windows Updates from reinstalling the packages
5. Finally, fix this issue, change system app removal to hard delete.

**Comment by undergroundwires** [comment]
Released in [0.12.6](https://github.com/undergroundwires/privacy.sexy/releases/tag/0.12.6) 🚀, thank you for the suggestion again @TheAndr0id, add a few "thank you"s to you in release message :)
